### ISSUE TITLE
Image flowtag not working anymore on firmware 2.x

### ISSUE THREAD

**Issue by jghaanstra** [athombv/apps-sdk-issues#49]
Two of my apps (DoorBird and Android IP Webcam) use the image flowtag functionality as documented here: https://apps.developer.athom.com/tutorial-Images.html

This worked reasonably well on firmware 1.x although there was an issue with registering multiple image flowtags from the same app. With firmware 2.x this functionality no longer works at all. My code is conform the documentation but I still receive an error somewhere during registering the image token indicating there is something wrong in the core code. The error is as following.

```
2019-02-06 19:31:16 [log] [DoorbirdApp] Initializing DoorBird app ...
(node:24624) UnhandledPromiseRejectionWarning: Error: data is not defined
    at Object.JSON.parse (/opt/homey-client/system/helpers/jsonfns.js:1:1190)
    at JSON.parse (<anonymous>)
    at jsonToObject (/opt/homey-client/system/helpers/jsonfns.js:1:1083)
    at function.e.data.map.e (/opt/homey-client/system/manager/ManagerApps/bootstrap/sdk/v2/lib/HomeyClient.js:1:1379)
    at Array.map (<anonymous>)
    at HomeyClient._onMessage (/opt/homey-client/system/manager/ManagerApps/bootstrap/sdk/v2/lib/HomeyClient.js:1:1372)
    at emitTwo (events.js:126:13)
    at process.emit (events.js:214:7)
    at emit (internal/child_process.js:762:12)
    at _combinedTickCallback (internal/process/next_tick.js:142:11)
(node:24624) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
``` 

Code can be found here: https://github.com/jghaanstra/com.doorbird/blob/master/app.js

**Comment by WeeJeWel** [comment]
Could you simplify your code to a reproducible piece?

**Comment by jghaanstra** [comment]
I see I linked to the wrong file. But here is the code related to the issue.

```
    // LIVE SNAPSHOT TOKEN
    let doorbirdSnapShot = new Homey.Image('jpg');
    doorbirdSnapShot.setBuffer(async () => {
      try {
        const image = await util.createSnapshot(this.getSetting('address'), this.getSetting('username'), this.getSetting('password'))
        if (image) {
          return image;
        } else {
          throw new Error('Invalid Response');
        }
      } catch(error) {
        this.error(error);
        throw new Error('Invalid Response');
      }
    });

    doorbirdSnapShot.register()
      .then(() => {
        let doorbirdSnapShotToken = new Homey.FlowToken('doorbird_snapshot', {
          type: 'image',
          title: Homey.__('Live Snapshot')
        })

        doorbirdSnapShotToken
          .register()
          .then(() => {
            doorbirdSnapShotToken.setValue(doorbirdSnapShot);
          })
          .catch(this.error.bind(this, 'doorbirdSnapShotToken.register'));

      })
      .catch(this.error.bind(this, 'doorbirdSnapShot.register'));
```

**Comment by WeeJeWel** [comment]
Thank you. After executing which piece are you getting the error?

**Comment by jghaanstra** [comment]
The error pops up when starting the app, I assume when it tries to register the image token.

**Comment by WeeJeWel** [comment]
It would help a great deal if you could isolate the issue to a specific call!

**Comment by jghaanstra** [comment]
Yep, it happens when registering the image token. In my code example in `doorbirdSnapShotToken     .register()` . The promise is not resolved or rejected.

**Comment by WeeJeWel** [comment]
Ah I see. You don't need to create a Homey.FlowToken, you can simply pass the `Image` instance in the `tokens` object in `FlowCardTrigger.trigger()`

**Comment by WeeJeWel** [comment]
Best practice:

* Create one image and register it that always returns the latest snapshot
* Update this image when you know it has changed, and:
* Reference this image when triggering a Flow

**Comment by jghaanstra** [comment]
Ok, thnx. You might want to update the docs for this as well: https://apps.developer.athom.com/tutorial-Flow-Tokens-Image.html

**Comment by WeeJeWel** [comment]
Hm sorry, I think I was too fast — you weren't triggering a Flow. I'll try if the example works and if not fix it.

**Comment by jghaanstra** [comment]
I just tested this with the [Image Grabber app](https://github.com/jordenc/image.grabber) from Jorden Chamid. It has the same issue as my DoorBird and Android IP Webcam app but might be easier to use reproduce it.

Just grab any snapshot from a camera with the Image Grabber app and try to send it to Slack using the send image card. Be sure to grab the global token which is registered by the Image Grabber app and use that in the Slack action card. It will then cause Slack to crash because the image buffer for the global token is null and return the error from this issue in the Image Grabber app similar to how this happens on my DoorBird and Android IP Webcam app.

**Comment by jordenc** [comment]
Only works when using the local tag, indeed, otherwise the image is null indeed. I receive mails about the same error very often with Telegram and email-sender apps.

**Comment by jordenc** [comment]
A "PS" from my side. I always create a new image token in my apps, which I probably shouldn't do. When I go to https://developer.athom.com/ and login, under "Tools -> Images", I see multiple images.

I probably should just create a new image token in the start of the device or app (depends on what kind of app it is... for a Synology webcam I want one image per device, in ImageGrabber I should just use 1 image token in total I think). However, I also have the Netatmo app from Athom which also makes multiple images. My Homey is getting sluggish last few days and I think it has to do with too many images in memory.
Besides that, also some Sonos images in memory.

**Comment by jghaanstra** [comment]
@jorden You should register a global image token for each image token you wish to use. The image buffer from a global image token is updated everytime it's requested. This means Homey only keeps one image buffer in memory per registered image token. An app can ofcourse have multiple global image tokens for different purposes. For instance, in my DoorBird app I register three global image tokens: a live snapshot, a history snapshot from latest motion event and a history snapshot from latest doorbell event. These three show up as three images under the developer tools. But currently they are not working there either and just show an `invalid_image_buffer` because the image tokens have not been populated correctly due to this issue.

@WeeJeWel is there any progress on this?

**Comment by WeeJeWel** [comment]
This has been fixed in Homey v2.0.5. Thank you for reporting the issue!
